exonum/src/encoding/protobuf/mod.rs

```diff
@@ -0,0 +1,247 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![allow(bare_trait_objects)]
+
+//! Protobuf generated structs
+//!
+mod blockchain;
+mod helpers;
+mod protocol;
+
+pub use self::blockchain::{Block, ConfigReference, TxLocation};
+pub use self::helpers::{BitVec, Hash, PublicKey};
+pub use self::protocol::{
+    BlockRequest, BlockResponse, Connect, PeersRequest, Precommit, Prevote, PrevotesRequest,
+    Propose, ProposeRequest, Status, TransactionsRequest, TransactionsResponse,
+};
+
+use chrono::{DateTime, TimeZone, Utc};
+use crypto;
+use encoding::Error;
+use helpers::{Height, Round, ValidatorId};
+use messages::BinaryForm;
+use protobuf::{well_known_types, Message, RepeatedField};
+
+/// Used for establishing correspondence between rust struct
+/// and protobuf rust struct
+pub trait ToProtobuf: Sized {
+    /// Type of the protobuf clone of Self
+    type ProtoStruct: Message;
+
+    /// Struct -> ProtoStruct
+    fn to_pb(&self) -> Self::ProtoStruct;
+
+    /// ProtoStruct -> Struct
+    fn from_pb(pb: Self::ProtoStruct) -> Result<Self, ()>;
+}
+
+impl<T> BinaryForm for T
+where
+    T: ToProtobuf,
+{
+    fn encode(&self) -> Result<Vec<u8>, Error> {
+        Ok(self.to_pb().write_to_bytes().unwrap())
+    }
+
+    fn decode(buffer: &[u8]) -> Result<Self, Error> {
+        let mut pb = <Self as ToProtobuf>::ProtoStruct::new();
+        pb.merge_from_bytes(buffer).unwrap();
+        Ok(Self::from_pb(pb).unwrap())
+    }
+}
+
+/// Used for establishing correspondence between rust struct field
+/// and protobuf value
+pub trait ProtobufValue: Sized {
+    /// Type of value that is returned with pb.take_field() method
+    type ProtoValue;
+    /// RustField -> ProtobufRustField
+    fn to_pb_field(&self) -> Self::ProtoValue;
+    /// ProtobufRustField -> RustField
+    fn from_pb_field(pb: Self::ProtoValue) -> Result<Self, ()>;
+}
+
+impl<T> ProtobufValue for T
+where
+    T: ToProtobuf,
+{
+    type ProtoValue = <Self as ToProtobuf>::ProtoStruct;
+
+    fn to_pb_field(&self) -> Self::ProtoValue {
+        self.to_pb()
+    }
+
+    fn from_pb_field(pb: Self::ProtoValue) -> Result<Self, ()> {
+        Self::from_pb(pb)
+    }
+}
+
+impl ToProtobuf for crypto::Hash {
+    type ProtoStruct = Hash;
+
+    fn to_pb(&self) -> Hash {
+        let mut hash = Hash::new();
+        hash.set_data(self.as_ref().to_vec());
+        hash
+    }
+
+    fn from_pb(pb: Hash) -> Result<Self, ()> {
+        let data = pb.get_data();
+        if data.len() == crypto::PUBLIC_KEY_LENGTH {
```

Comment:
Note that `HASH_SIZE = PUBLIC_KEY_LENGTH` may not hold for every crypto-backend (e.g. GOST).
Please consider changing this to `crypto::HASH_SIZE`.
```suggestion
        if data.len() == crypto::HASH_SIZE {
```

Reply:
Thanks, it was copy-paste error.